File: 30_days_of_code/20-sorting/elixir/sorting.ex
defmodule Solution do
  import Enum
  import List

  def bubble([h1 | [h2 | tail]], state) do
    if h1 > h2 do
      {arr, state} = bubble([h1 | tail], state + 1)
      {[h2 | arr], state}
    else
      {arr, state} = bubble([h2 | tail], state)
      {[h1 | arr], state}
    end
  end

  def bubble([h | []], state) do
    {[h], state}
  end

  def loop(list, state) do
    {list_copy, state} = bubble(list, state)
    if list_copy != list do
      {list_copy, state} = loop(list_copy, state)
    end
    {list_copy, state}
  end

  def input() do
    IO.gets("")
    list = map(String.split(IO.gets("")), fn(x) ->
      {n, _} = Integer.parse(x)
      n
    end)

    {sorted, state} = loop(list, 0)
    IO.puts "Array is sorted in #{state} swaps."
    IO.puts "First Element: #{first(sorted)}"
    IO.puts "Last Element: #{last(sorted)}"
  end
end

Solution.input

File: 30_days_of_code/20-sorting/cpp/sorting.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

void make_swap(vector<int>::iterator a, vector<int>::iterator b) {
  int t = *a;
  *a = *b;
  *b = t;
};

int main() {
  int n, count = 0;
  vector<int> arr;
  bool continue_sort;

  cin >> n;

  arr = vector<int>(n);

  for(int i = 0; i < n; i++) {
    cin >> arr[i];
  }

  do {
    continue_sort = false;
    for(vector<int>::iterator it = arr.begin(); it != arr.end() - 1; it++) {
      if(*it > *(it + 1)) {
        make_swap(it, it+1);
        count++;
        continue_sort = true;
      }
    }
  } while(continue_sort);

  cout << "Array is sorted in " << count << " swaps." << endl;
  cout << "First Element: " << *arr.begin() << endl;
  cout << "Last Element: " << *(arr.end()-1) << endl;

  return 0;
}
